### PR TITLE
Implement ZeroCopyInputStream for file section

### PR DESCRIPTION
**Pull request by dimitry-**
This implementation is intended to be used for reading protos from
capture file sections.

Test: run CaptureFileTests

### CODE REVIEW

File: src/CaptureFile/FileFragmentInputStream.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include <google/protobuf/io/zero_copy_stream.h>
+
+#include <optional>
+
+#include "OrbitBase/File.h"
+
+namespace orbit_capture_file_internal {
+
+// ZeroCopyInputStream implementation for a file fragment with offset and size.
+// this class is used to read protos from capture file sections and makes sure
```
Comment (florian-kuebler) [inline src/CaptureFile/FileFragmentInputStream.h:17]:
```suggestion
// This class is used to read protos from capture file sections and makes sure
```

File: src/CaptureFile/FileFragmentInputStream.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                                   uint64_t size, size_t block_size = 1 << 16)
+      : fd_{fd},
+        file_fragments_start_{file_offset},
+        file_fragments_end_{file_offset + size},
+        buffer_(block_size),
+        current_position_{file_offset} {
+    CHECK(size > 0);
+  }
+
+  bool Next(const void** data, int* size) override;
```
Comment (florian-kuebler) [inline src/CaptureFile/FileFragmentInputStream.h:31]: Can you add a super brief comment to at least the first two methods?

File: src/CaptureFile/FileFragmentInputStream.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  current_position_ += bytes_read;
+  (*data) = buffer_.data();
+  (*size) = static_cast<int>(bytes_read);
+
+  return true;
+}
+
+void FileFragmentInputStream::BackUp(int count) {
+  CHECK(count >= 0);
+  if (static_cast<size_t>(count) > current_position_) {
```
Comment (florian-kuebler) [inline src/CaptureFile/FileFragmentInputStream.cpp:42]: could you modify the type of count instead of casting?

Comment (florian-kuebler) [reply src/CaptureFile/FileFragmentInputStream.cpp:42]: Also, I have the feeling that a comment above this if would help.

File: src/CaptureFile/FileFragmentInputStream.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    current_position_ = file_fragments_start_;
+    return;
+  }
+
+  current_position_ = std::max(file_fragments_start_, current_position_ - count);
+}
+
+bool FileFragmentInputStream::Skip(int count) {
+  CHECK(count >= 0);
```
Comment (florian-kuebler) [inline src/CaptureFile/FileFragmentInputStream.cpp:51]: Again, could you modify the type of count instead?

File: src/CaptureFile/FileFragmentInputStreamTest.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  // Make sure we do not go out of fragment boundary
+  input_stream.BackUp(100);
+  EXPECT_EQ(input_stream.ByteCount(), 0);
+
+  ASSERT_TRUE(input_stream.Next(&bytes, &size));
+  EXPECT_EQ((std::string_view{static_cast<const char*>(bytes), static_cast<size_t>(size)}),
+            "urna moles");
+  EXPECT_EQ(input_stream.ByteCount(), 10);
+
+  // do not read past end of the fragment
```
Comment (florian-kuebler) [inline src/CaptureFile/FileFragmentInputStreamTest.cpp:48]:
```suggestion
  // Do not read past end of the fragment
```

File: src/CaptureFile/CMakeLists.txt [hunk 6]
```diff
@@ -19,7 +19,9 @@ target_sources(
   CaptureFile
   PRIVATE CaptureFileConstants.h
           CaptureFile.cpp
-          CaptureFileOutputStream.cpp)
+          CaptureFileOutputStream.cpp
+          FileFragmentInputStream.cpp
```
Comment (florian-kuebler) [inline src/CaptureFile/CMakeLists.txt:23]: This class feels rather generic to me, should the be inside of some other module (I have no strong opinion on this though).

File: src/CaptureFile/FileFragmentInputStream.cpp (hunk 3 above)
Comment (dimitry-) [reply src/CaptureFile/FileFragmentInputStream.cpp:42]: well, the interface is set by google::protobuf::io::ZeroCopyInputStream, I can't change the type here.

File: src/CaptureFile/FileFragmentInputStream.cpp (hunk 4 above)
Comment (dimitry-) [reply src/CaptureFile/FileFragmentInputStream.cpp:51]: same reason.

File: src/CaptureFile/FileFragmentInputStream.cpp (hunk 3 above)
Comment (florian-kuebler) [reply src/CaptureFile/FileFragmentInputStream.cpp:42]: Okay, then never mind.

File: src/CaptureFile/CMakeLists.txt (hunk 6 above)
Comment (dimitry-) [reply src/CaptureFile/CMakeLists.txt:23]: I do not expect it to be used anywhere else.

File: src/CaptureFile/FileFragmentInputStream.h (hunk 1 above)
Comment (dimitry-) [reply src/CaptureFile/FileFragmentInputStream.h:17]: done.

File: src/CaptureFile/FileFragmentInputStream.h (hunk 2 above)
Comment (dimitry-) [reply src/CaptureFile/FileFragmentInputStream.h:31]: added brief comments and links

File: src/CaptureFile/CMakeLists.txt (hunk 6 above)
Comment (florian-kuebler) [reply src/CaptureFile/CMakeLists.txt:23]: Okay, thanks. That is fine for me.

File: src/CaptureFile/FileFragmentInputStream.h (hunk 2 above)
Comment (florian-kuebler) [reply src/CaptureFile/FileFragmentInputStream.h:34]: Thanks!

**Review by florian-kuebler — APPROVED** [review]
Thanks for addressing my comments. LGTM!

**Review by antonrohr — APPROVED** [review]
LGTM. Nit: Maybe one more test for BackUp, see:

https://storage.googleapis.com/orbitprofiler-build/unit-test-coverage/prod/orbitprofiler/builds/linux/coverage_clang9/presubmit/1221/20210503-014658/coverage_CaptureFile/coverage/mnt/github/orbitprofiler/src/CaptureFile/FileFragmentInputStream.cpp.html

**Comment by dimitry-** [comment]
> LGTM. Nit: Maybe one more test for BackUp, see:
> 
> https://storage.googleapis.com/orbitprofiler-build/unit-test-coverage/prod/orbitprofiler/builds/linux/coverage_clang9/presubmit/1221/20210503-014658/coverage_CaptureFile/coverage/mnt/github/orbitprofiler/src/CaptureFile/FileFragmentInputStream.cpp.html

done

**Pull request merged**